rand_distr/src/poisson.rs

```diff
@@ -145,11 +143,21 @@ where
         let method = if lambda < F::from(12.0).unwrap() {
             Method::Knuth(KnuthMethod::new(lambda))
         } else {
+            if lambda > F::from(Self::MAX_LAMBDA_POISSON).unwrap() {
+                return Err(Error::ShapeTooLarge);
+            }
             Method::Rejection(RejectionMethod::new(lambda))
         };
 
         Ok(Poisson(method))
     }
+
+    /// Maximum value for `lambda` in the Poisson distribution.
+    /// This will make sure the samples will fit in a `u64`.
+    /// It also makes sure we do not run into numerical problems with very large `lambda`.
+    /// `1.844e19 + 1_000_000 * sqrt(1.844e19) < 2^64 - 1`,
+    /// so the probability of getting a value larger than `u64::MAX` is << 1e-1000.
+    pub const MAX_LAMBDA_POISSON: f64 = 1.844e19;
```

Comment:
Rename to just `MAX_LAMBDA` since this is an associated constant of `Poisson`.

I'll edit the doc for you:
```
    /// The maximum supported value of `lambda`
    ///
    /// This value was selected such that
    /// `MAX_LAMBDA + 1e6 * sqrt(MAX_LAMBDA) < 2^64 - 1`,
    /// thus ensuring that the probability of sampling a value larger than
    /// `u64::MAX` is less than 1e-1000.
    ///
    /// Applying this limit also solves
    /// [#1312](https://github.com/rust-random/rand/issues/1312).
```